src/pipecat/services/aws_nova_sonic/aws.py

```diff
@@ -0,0 +1,973 @@
+        self._prompt_name = None
+        self._input_audio_content_name = None
+        self._content_being_received = None
+        self._assistant_is_responding = False
+        self._context_available = False
+        self._ready_to_send_context = False
+        self._handling_bot_stopped_speaking = False
+        self._triggering_assistant_response = False
+        self._assistant_response_trigger_audio: bytes = None  # Not cleared on _disconnect()
+        self._disconnecting = False
+        self._connected_time: float = None
+
+    #
+    # standard AIService frame handling
+    #
+
+    async def start(self, frame: StartFrame):
+        await super().start(frame)
+        # TODO: maybe connect but don't send history until we get all of our settings?
+        # how do we know how long to wait?
+        # ah, i think we'll *always* get at least one OpenAILLMContextFrame which kicks things off
+        # so we need to send the initial history when:
+        # - we're connected
+        # - we've gotten the first context
+        # i *think* this is what's controlled by _api_session_ready/_run_llm_when_api_session_ready
+        await self._start_connecting()
+
+    async def stop(self, frame: EndFrame):
+        await super().stop(frame)
+        await self._disconnect()
+
+    async def cancel(self, frame: CancelFrame):
+        await super().cancel(frame)
+        await self._disconnect()
+
+    #
+    # conversation resetting
+    #
+
+    async def reset_conversation(self):
+        logger.debug("Resetting conversation")
+        await self._disconnect()
+        await self._start_connecting()
+        # Use existing context
+        self._context_available = True
+        await self._finish_connecting_if_context_available()
+
+    #
+    # frame processing
+    #
+
+    async def process_frame(self, frame: Frame, direction: FrameDirection):
+        await super().process_frame(frame, direction)
+
+        if isinstance(frame, OpenAILLMContextFrame):
+            await self._handle_context(frame.context)
+        elif isinstance(frame, InputAudioRawFrame):
+            await self._handle_input_audio_frame(frame)
+        elif isinstance(frame, BotStoppedSpeakingFrame):
+            await self._handle_bot_stopped_speaking()
+        elif isinstance(frame, AWSNovaSonicFunctionCallResultFrame):
+            await self._handle_function_call_result(frame)
+        # TODO: do we need to do anything for the below four frame types?
+        elif isinstance(frame, StartInterruptionFrame):
+            # print("[pk] StartInterruptionFrame")
+            pass
+        elif isinstance(frame, UserStartedSpeakingFrame):
+            # print("[pk] UserStartedSpeakingFrame")
+            pass
+        elif isinstance(frame, StopInterruptionFrame):
+            # print("[pk] StopInterruptionFrame")
+            pass
+        elif isinstance(frame, UserStoppedSpeakingFrame):
+            # print("[pk] UserStoppedSpeakingFrame")
+            pass
```

Comment:
Inside the GeminiMultimodalLiveLLMService, it looks like we use the UserStartedSpeakingFrame and UserStoppedSpeakingFrame to buffer the user audio and later transcribe it. But I believe we're already receiving the user transcription here, right?

As for interruptions, my guess is that they're automatically handled by Nova Sonic as well.

So, I don't think we need to handle those frames here.

Reply:
Wonder if we could use the `UserStoppedSpeaking` frame to better aggregate the user transcriptions. As it is today, they can end up split up into separate messages in the context. Anyway, jotting this down as a later TODO.